File: exercises/exercise-1-testing-java/src/main/java/io/miragon/example/application/service/bike/CheckBikeAvailabilityService.java
package io.miragon.example.application.service.bike;

import io.miragon.example.application.port.inbound.bike.CheckBikeAvailabilityUseCase;
import io.miragon.example.application.port.outbound.bike.BikeSubscriptionRepository;
import io.miragon.example.domain.bike.BikeSubscriptionId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.Random;

@Service
public class CheckBikeAvailabilityService implements CheckBikeAvailabilityUseCase {

    private static final Logger log = LoggerFactory.getLogger(CheckBikeAvailabilityService.class);

    private final BikeSubscriptionRepository repository;
    private final Random random = new Random();

    public CheckBikeAvailabilityService(BikeSubscriptionRepository repository) {
        this.repository = repository;
    }

    @Override
    public boolean checkAvailability(BikeSubscriptionId subscriptionId) {
        var subscription = repository.find(subscriptionId);

        // Mock availability check: 80% chance of being available
        boolean available = random.nextDouble() < 0.8;

        log.info("Bike {} availability check: {}", subscription.bikeId().value(), available);

        repository.save(subscription.markAvailability(available));

        return available;
    }

}

File: exercises/exercise-1-testing-kotlin/src/main/kotlin/io/miragon/example/application/service/bike/CheckBikeAvailabilityService.kt
package io.miragon.example.application.service.bike

import io.miragon.example.application.port.inbound.bike.CheckBikeAvailabilityUseCase
import io.miragon.example.application.port.outbound.bike.BikeSubscriptionRepository
import io.miragon.example.domain.bike.BikeSubscriptionId
import mu.KotlinLogging
import org.springframework.stereotype.Service
import kotlin.random.Random

@Service
class CheckBikeAvailabilityService(
    private val repository: BikeSubscriptionRepository
) : CheckBikeAvailabilityUseCase {

    private val log = KotlinLogging.logger {}

    override fun checkAvailability(subscriptionId: BikeSubscriptionId): Boolean {
        val subscription = repository.find(subscriptionId)

        // Mock availability check: 80% chance of being available
        val available = Random.nextDouble() < 0.8

        log.info { "Bike ${subscription.bikeId.value} availability check: $available" }

        repository.save(subscription.markAvailability(available))

        return available
    }
}
